R-packages/modeltools/R/deriv.R

```diff
@@ -0,0 +1,205 @@
+#'   `df` should be one of "min" or "1se" indicating the selection rule to use
+#'   based on the cross-validation error curve (minimum or 1-standard-error
+#'   rule, respectively). Default is 8 when `cv = FALSE`, and "1se" when `cv =
+#'   TRUE`.} 
+#' }
+#' 
+#' @return A data frame given by appending a new column to `x` named according
+#'   to the `col_name` argument, containing the estimated derivative values. 
+#'
+#' @export
+estimate_deriv = function(x, method = c("lin", "ss", "tf"), n = 14,
+                          col_name = "deriv", keep_obj = FALSE, deriv = 1,
+                          ...) {  
+  # Define the slider function
+  method = match.arg(method)
+  slide_fun = switch(method,
+                     "lin" = linear_reg_deriv,
+                     "ss" = smooth_spline_deriv,
+                     "tf" = trend_filter_deriv)
+  
+  # Check the derivative order
+  if (!(deriv == 1 || deriv == 2)) {
+    stop("`deriv` must be either 1 or 2.")
+  }
+
+  # Slide the derivative function
+  x = slide_by_geo(x, slide_fun, n, col_name = "temp", col_type = "list",
+                   keep_obj = keep_obj, deriv = deriv, ...)
+
+  # Grab the derivative result
+  x = x %>% rowwise() %>% mutate(!!col_name := temp$result)
+
+  # Grab the derivative object, if we're asked to
+  if (keep_obj) {
+    x = x %>% rowwise() %>%
+      mutate(!!paste0(col_name, "_obj") := list(temp$object))
+  }
+  
+  # Delete the temp column and return
+  return(select(x, -temp))
+}
+
+#' Compute derivatives function via linear regression
+#' @importFrom stats lsfit coef
+#' @importFrom tidyr drop_na
+#' @noRd
+linear_reg_deriv = function(x, ...) {
+  params = list(...)
+  params[[1]] = NULL # dplyr::group_modify() includes the group here
+
+  # If derivative order is 2, then return trivial result
+  if (params$deriv == 2) return(list(object = NULL, result = 0))
+  
+  keep_obj = params$keep_obj
+  params$keep_obj = NULL
+  params$deriv = NULL
+  x = x %>% select(time_value, value) %>% drop_na()
+  params$x = as.numeric(x$time_value)
+  params$y = x$value
+  
+  return(tryCatch(suppressWarnings(suppressMessages({
+    object = do.call(lsfit, params)
+    result = End(coef(object))
+    if (!keep_obj) object = NULL # For memory sake
+    list(object = object, result = result)
+  })),
+  error = function(e) list(object = NA, result = NA)))
+}
+
+#' Compute derivatives function via smoothing spline
+#' @importFrom stats smooth.spline predict
+#' @noRd
+smooth_spline_deriv = function(x, ...) {
+  params = list(...)
+  params[[1]] = NULL # dplyr::group_modify() includes the group here
+
+  keep_obj = params$keep_obj
+  deriv = params$deriv
+  params$keep_obj = NULL
+  params$deriv = NULL
+  x = x %>% select(time_value, value) %>% drop_na()
+  params$x = as.numeric(x$time_value)
+  params$y = x$value
+  
+  return(tryCatch(suppressWarnings(suppressMessages({
+    object = do.call(smooth.spline, params)
```

Comment:
```suggestion
    capture.output(object <- do.call(smooth.spline, params))
```
Diverts anything written by `smooth.spline` away from stdout.